Add Package Control docs (#134)
sublimetext-io/docs.sublimetext.io: docs/reference/package-control/repository.md

---
title: "Package Control: repository.json"
---

<!-- Original: -->
<!-- https://github.com/wbond/package_control/blob/master/example-repository.json -->

# Examples for the repository.json

The default `repository.json` is managed in the [Package Control Default Channel][pcc_json] repo.

[pcc_json]: https://github.com/wbond/package_control_channel/blob/master/repository.json

## Properties

- `$schema`
- `schema_version`
- `packages`

> - `libraries`

Reviewer: In schema v3, this is called "dependencies".  "libraries" are schema v4.

Reviewer: All keys except schema_version are optional btw.

Author: > The major question is, which schema version to describe.

I think they should cover v4 where it makes sense, which is for the channel and the libraries. The section about packages should contain information developers of packages can actually use, and should thus cover v3. That is also the only documentation I'm able to write. If the libraries section didn't already exist I wouldn't have put it in this PR: I know too little about it to effectively write about it. 

So, perhaps technically it's a bit weird that there is a libraries section (which is v4 schema) in a document mostly concerning v3 schema. Perhaps the entire libraries section should be moved to a separate doc. I welcome someone who understands that section to do so, perhaps as a PR on top of this one? But it matches the current status quo and it's the most useful documentation for each main audience (package authors, and library authors).

- `includes`

`$schema` is used by [LSP][lsp] and [LSP-json][lspjson]
to provide linting and completions.

> `schema_version` is currently at [3.0.0][schema].

Reviewer 2: ```suggestion
`schema_version` of default channel is currently restricted to [3.0.0][schema]
for backward compatibility reasons with packagecontrol.io.
```

```json
{
    "$schema": "sublime://packagecontrol.io/schemas/repository",
    "schema_version": "3.0.0",
    "packages": [],
    "libraries": [],
    "includes": []
}
```

[lsp]: https://packages.sublimetext.io/packages/LSP/
[lspjson]: https://packages.sublimetext.io/packages/LSP-json/
[schema]: https://github.com/wbond/package_control/blob/master/sublime-package.json#L505

## `includes`

If you want to split your repository up into multiple smaller files
for the sake of organization,
the `includes` key allows you to enter URL paths
that will be combined together.
These URLs these can be relative or absolute.
This is also how the default repository is managed.

All repositories must be an HTTPS URL, or a local file path.

The following values are supported:

- an external repository.json URL
- a repository located relative to this one, via a relative path
- a `file:///` protocol URL

```json
{
	"includes": [
		"./repository/0-9.json",
		"./repository/a.json",
		"file:///absolute/path/to/repository.json",
		"https://packagecontrol.io/repository/b.json"
	]
}
```

## `packages`

> The only required properties for each package are:

Reviewer: ```suggestion
For open source packages, the only required properties for each package are:
```

Author: The word "required" in the context of schemas has a specific meaning. Here probably something else was meant: when submitting a package to the default channel releases and details are required. I rephrased this section to hopefully be a bit clearer about this.

Reviewer: We could say: required are name, author and releases.  name plus author are typically computed from the details. And releases are typically dynamic, e.g. tags: true or branch: main.  Or static, providing url, version, date, sublime_text.  (Note: the crawler turns dynamic release sections into static ones.  I.e. it resolves branch: main to a static version/url.)

(We could make releases optional in the future and default to  sublime_text: all, tags: true.  That would make our public registry shorter.  It is generally good to have defaults for fields.)

Author: > We could make releases optional in the future and default to sublime_text: all, tags: true.

I like that, that would work for most packages.

Reviewer: Completely off topic, but I noticed you changed "authors" to "author" although it is a `List[str]`.  Why ist that, if we also have the plural form in "releases", "labels"?

Reviewer 2: I meant one of "asset", "branch" or "tag" for dynamic release templates.

A static/resolved release requires "url", "version" and "date", optionally "sha256".

The `date` part is used to calculate `last_modified` if it is not already given explicitly. That's at least original JsonRepositoryProvider behavior. 

Is `date` crucial, hard to say as dynamic releases resolve to a static release with exactly these fields.

IMHO, usage of static releases should be reduced to a minimum as it always requires them to be updated for each release, which is one reason for the huge amount of external repos.

If a build-step is needed, and github releases can be used for deployments, asset based releases are the alternative of choice.

Static releases should ideally only be the result/output of a crawler run.

Reviewer 2: > Completely off topic, but I noticed you changed "authors" to "author" although it is a `List[str]`. Why ist that, if we also have the plural form in "releases", "labels"?

Because `"author"` is what is historically used by all dynamic repositories and channels, and "authors" was just present in channel_v3.json generated by packagecontrol.io crawler as it carelessly just mapped internal database fields.

Hence I made a decision for one of them, and that was the more proimenent one.

Reviewer 2: > I think you dropped the requirement of date in PC. Maybe check that.

Not entirely as it seems. At least JsonRepositoryProvider's get_packages() method relies on thosse being present to calculate `last_modified`. Dropping it requires an overhaul of the repo clients by merging Client and Provider logic.

Reviewer: You make the version test in https://github.com/wbond/package_control/blob/6cb331ed514802e342ede5356d836508a03d5166/package_control/providers/json_repository_provider.py#L680-L697 very obvious but I don't see the "date" check, and when computing "last_modified" you defensively use `get("date")` which was a `["date"]` before.  I don't think we should drop "date" but I noticed that slight variation.  But maybe you moved the check to a different place?

				{
					"sublime_text": "*",
					"tags": true
				}
			]
		}
	]
}
```

This is more complete, and adds
a specific name, labels and homepage and donation links.

```json
{
	"packages": [
		{
			"name": "Solarized Color Scheme",
			"details": "https://github.com/braver/Solarized",
			"homepage": "https://ethanschoonover.com/solarized/",
			"donate": "https://paypal.me/koenlageveen",
			"labels": ["color scheme"],
			"releases": [
				{
					"sublime_text": ">=3000",
					"tags": true
				}
			]
		}
	]
}
```

You can find tons of examples in the default repository
at https://github.com/wbond/package_control_channel/tree/master/repository.

### Package metadata without `details`

Package metadata are resolved using `details`,
where each field can be overridden by explicitly set values.
If `details` is absent, each release must specify a `base` url.
In this and other aspects,
release based packages behave the same as tag based packages.

```json
{
	"packages": [
		{
			"name": "Package With Asset",
			"description": "A library with custom download assets",
			"author": "Jon Doe",
			"releases": [
				{
					"base": "https://github.com/SublimeText/PackageWithAsset",

> 					"asset": "Package.With.Asset-${version}-st${st_build}-${platform}.sublime-package",

Reviewer: "asset" is actually not supported by neither the new nor the old crawler.

Reviewer 2: It is part of schema 4.0.0 and must be supported, however. Omitting it here is the wrong direction.

Author: There are a couple of features in here that are not supported or simply not allowed, on the public default channel. It makes sense the crawler for that channel doesn’t support the entire schema. I’ll add a note/caveat about the scope of this feature. Let me know if you spot any others.

Reviewer 2: I disagree. The crawler has to fully support v4 schema to make use of all of its features as soon as we switch to the new site. That's the main point of all of this. Otherwise we can stop here.

Author: There is a good chance I’m misunderstanding things, so sorry for that. It’s my understanding that the crawler exclusively crawls the public default channel. We disallow certain constructions there, so the crawler doesn’t need to support them right? Or, hm, now that I think about it, probably some of the repositories included in the channel _can_ use those features… so even if those are edge cases or currently not in use, the crawler should support them… hmmmm!

Anyway, this is not a discussion about the schema or the design of the crawler. I don’t have opinions on either. The document should document the schema primarily of course. My intention was mostly to add something like this:
“If you’re looking to add a package to the default channel, you would not use assets.”.

a `url` to the zip, and a `date`.
Because this requires an update of the repository file for each release
it is not allowed in the default repository.

Versions needs to be numbered in accordance with [semver](https://semver.org). 

The URL needs to be a zip file containing the package.
It is permissible for the zip file to contain a single root folder
with any name.
All files will be extracted out of this single root folder.
This allows zip files from GitHub, GitLab, and Bitbucket to be used as sources.
The URL can be a relative path
from the location of the repository.json file that specifies it.

The date must be in the form "YYYY-MM-DD HH:MM:SS" and should be UTC.

```json
{
	"packages": [
		{
			"details": "https://github.com/wbond/sublime_alignment",
			"releases": [
				{
					"version": "2.0.0",
					"url": "https://codeload.github.com/wbond/sublime_alignment/zip/v2.0.0",
					"date": "2011-09-18 20:12:41",
					"sublime_text": "*"
				}
			]
		},
		{
			"details": "../alignment",
			"releases": [
				{
					"version": "2.0.0",
					"url": "./downloads/alignment-2.0.0.zip",
					"date": "2011-09-18 20:12:41",
					"sublime_text": "*"
				}
			]
		}
	]
}
```

> ### `asset`

Reviewer: As above, "asset" is not supported for packages but for libraries.  Only deathaxe's crawler actually evaluates these template strings.

an explicit release specification can be declared for each platform.

```json
{
	"details": "https://github.com/SublimeText/PackageWithAsset",
	"releases": [
		{
			"asset": "FileName-linux-aarch64.sublime-package",
			"platforms": ["linux-arm64"]
		},
		{
			"asset": "FileName-linux-amd64.sublime-package",
			"platforms": ["linux-x64"]
		},
		{
			"asset": "FileName-macosx-amd64.sublime-package",
			"platforms": ["osx-arm64"]
		},
		{
			"asset": "FileName-macosx-arm64.sublime-package",
			"platforms": ["osx-arm64"]
		},
		{
			"asset": "FileName-win-amd64.sublime-package",
			"platforms": ["windows-x64"]
		}
	]
}
```

## `libraries`

Note that the following documentation is not up to date
with latest developments in Sublime Text (ST) (e.g. the Python 3.8 plugin host)
or the Package Control 4.0 client package (PC4.0),
and is mostly included to document legacy use cases.

---

These are typically compiled Python extensions that are supplementary to,
or missing from Sublime Text.

Each library must have a name, description, author, issues URL,
and a list of releases.
Each release needs a `version` and `url` or `base` and `tags` keys.

> The sublime_text, platforms and python_versions keys are optional

Reviewer 3: ```suggestion
The `sublime_text`, `platforms`, and `python_versions` keys are optional
```

Sorry I missed this before. I had a question about libraries and went looking for the answer. :sweat_smile:

			"author": "wbond",
			"issues": "https://github.com/codexns/sublime-bz2/issues",
			"releases": [
				{

					"base": "https://github.com/codexns/sublime-bz2",
					"tags": true
				},
				{
					"base": "https://github.com/codexns/sublime-bz2",
					"tags": true,
					"platforms": ["*"],
					"python_versions": ["3.3"],
					"sublime_text": "*"
				}
			]
		},
		{
			"name": "ssl-linux",
			"description": "Python _ssl module for Linux",
			"author": "wbond",
			"issues": "https://github.com/codexns/sublime-ssl-linux/issues",
			"releases": [
				{
					"version": "1.0.0",
					"url": "http://packagecontrol.io/ssl-linux.sublime-package",
					"platforms": ["linux"],
					"sha256": "d12a2ca2843b3c06a834652e9827a29f88872bb31bd64230775f3dbe12e0ebd4"
				}
			]
		}
	]
}
```

An example legacy python 3.3 dependency for ST2 on Windows,
provided as explicit download asset secured by sha256 hash
as being shipped via HTTP.
Note that ST2 is no longer supported by PC4.0.

```json
{
	"name": "ssl-windows",
	"description": "Python _ssl module for Sublime Text 2 on Windows",
	"author": "wbond",
	"issues": "https://github.com/codexns/sublime-ssl-windows/issues",
	"releases": [
		{

> 

Reviewer 3: ```suggestion
```

## `packages`

For each package various properties can be specified,
most importantly `releases` and `details`.

- `releases`<br>
  Describes the [releases][] and their requirements.
- `details`<br>
  For packages hosted on GitHub, Bitbucket or GitLab,
  most properties of the package will be retrieved
  from the `details` URL.
  Note that for package submissions to the default repository,

>   `details` must always be provided.

Reviewer 3: I have no issue with `<br>`, but just in case you were unaware, you can get Markdown to add one with two trailing spaces. It can get annoying for editing and Git to have meaningful trailing whitespace, though. (And that might be why you used `<br>`s.)

Author: I am certainly aware and prefer double spaces, but the contribution guide told me not to:

>  Insert manual line breaks using <br> tags instead of using two trailing spaces.
